DOCSP-42334: Observables (#86)
mongodb/docs-scala: source/observables.txt

How to Process an Observable
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To run a MongoDB operation and process its data, you must request
the operation results from an ``Observable``. The ``Observable``
is a higher-level abstraction built on top of a cursor, or a 
mechanism that fetches results in batches. The {+driver-short+}
handles the cursor internally and exposes the operation
results through an observable interface.

Each operation returns an instance of the ``Observable[T]`` type, where the ``[T]``
parameterization corresponds to the type of each item stored by the
observable. Some operations return a specialized observable that
extends the general ``Observable[T]`` type and provides additional
methods to filter and process results. These specialized observables
include the following types:

- ``SingleObservable[T]``: Returned by operations that produce a single result
- ``FindObservable[T]``: Returned by the ``find()`` method
- ``DistinctObservable[T]``: Returned by the ``distinct()`` method
- ``AggregateObservable[T]``: Returned by the ``aggregate()`` method

.. tip::

  To view a full list of ``Observable`` types, see the `Observable <{+api+}/org/mongodb/scala/Observable.html>`__

>   API documentation.

Reviewer: Nit: We link to the Observable documentation a few other times on this page – can we remove this admonition?

- ``onError(e: Throwable)``: Called when the operation generates an error and prevents
  the ``Observer`` from receiving more data from the ``Observable``. You can define
  error handling logic by overriding this method.

- ``onComplete()``: Called when the ``Observer`` has consumed all results from the 
  ``Observable``. You can perform any final data processing by overriding this method.

The following sections show how to customize these methods to process read and
write operation results from an ``Observable``.

Access Read Operation Results
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To access data retrieved by a read operation, create an ``Observable[T]``
to store the operation results. Then, subscribe to the observable and
override the ``Observer`` class callback methods to process the results.

This example queries the ``restaurants`` database for documents
in which the ``cuisine`` value is ``"Czech"``. To retrieve
and process results, the example assigns a ``Observable[Document]``
to the operation and performs the following actions:

- Calls the ``subscribe()`` method to subscribe to the ``Observable``
  and passes an ``Observer`` as a parameter

> - Customizes the ``onNext()`` method to print each retrieved document,
>   which are ``Document`` instances
> - Customizes the ``onError()`` method to print any errors
> - Customizes the ``onComplete()`` methods to print a message after all
>   the results from the ``Observable`` are processed

Reviewer: Would using "Overrides" instead of "Customizes" here be more accurate/descriptive?

```suggestion
- Overrides the ``onNext()`` method to print each retrieved document,
  which are ``Document`` instances
- Overrides the ``onError()`` method to print any errors
- Overrides the ``onComplete()`` methods to print a message after all
  the results from the ``Observable`` are processed
```

      Iterable((_id,...), ..., (name,BsonString{value='Hospoda'}),
      (restaurant_id,BsonString{value='41569184'}))
      Iterable((_id,...), ..., (name,BsonString{value='Olde Prague Tavern'}),
      (restaurant_id,BsonString{value='41711983'}))
      Processed all results

Access Write Operation Results
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To access data retrieved by a write operation, create an ``Observable[T]``
to store the operation results. Then, subscribe to the observable and
override the ``Observer`` class callback methods to process the results.

This example inserts documents into the ``restaurants`` database
in which the ``cuisine`` value is ``"Nepalese"``. To retrieve
and process results, the example assigns an ``Observable[InsertManyResult]``
to the operation and performs the following actions:

- Calls the ``subscribe()`` method to subscribe to the ``Observable``
  and passes an ``Observer`` as a parameter
- Customizes the ``onNext()`` method to print the result of the insert
  operation, returned as an ``InsertManyResult``
- Customizes the ``onError()`` method to print any errors
- Customizes the ``onComplete()`` methods to print a message after all

>   the results from the ``Observable`` are processed

Reviewer: Same question/suggestion as above

.. io-code-block::
   :copyable:

   .. input:: /includes/observables.scala
      :language: scala
      :start-after: start-read-observable
      :end-before: end-read-observable
      :dedent:

   .. output::
      :language: console
      :visible: false

      AcknowledgedInsertManyResult{insertedIds={0=BsonObjectId{value=...},
      1=BsonObjectId{value=...}}}
      Processed all results

Use Lambda Functions to Process Results
---------------------------------------

Instead of explicitly overriding the callback functions from the
``Observer`` class, you can use lambda functions to concisely
process operation results. These functions allow you to use the
``=>`` arrow notation to customize the implementation of ``onNext()``,
``onError()``, and ``onComplete()``.

.. tip::

>    To learn more about lambda, or anonymous, functions,

Reviewer: S: This reads bit more smoothly for me but also if you have other ideas feel free to edit:

```suggestion
   To learn more about lambda functions, also known as anonymous functions,
```

   see the :wikipedia:`Anonymous function <w/index.php?title=Anonymous_function&oldid=1252038929>`
   Wikipedia entry.

Example
~~~~~~~

This example queries the ``restaurants`` database for each
distinct value of the ``borough`` field. The code subscribes
to the ``Observable`` returned by the ``distinct()`` method, then
uses lambda functions to print results and handle errors:

.. io-code-block::
   :copyable:

   .. input:: /includes/observables.scala
      :language: scala
      :start-after: start-lambda
      :end-before: end-lambda
      :dedent:

   .. output::
      :language: console
      :visible: false

      Bronx
      Brooklyn
      Manhattan
      Missing
      Queens
      Staten Island
      Processed all results

> 

Reviewer: Suggestion: I think we should also mention the `Await.result()` method as an alternative to the callbacks/lambdas mentioned on this page. See [Time Series](https://deploy-preview-86--docs-scala.netlify.app/databases-collections/time-series/) for examples, as this would probably warrant its own section on this page.

Author: Added a section about futures!

.. _scala-observables:

==============================
Access Data From an Observable
==============================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, subscribe, non-blocking

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to access the results of MongoDB operations
from an ``Observable`` instance.

An ``Observable`` represents a stream of data emitted by an operation
over time. To access this data, you can create an ``Observer`` instance
that subscribes to the ``Observable``. 

How to Process an Observable
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To run a MongoDB operation and process its data, you must request

> the operation results from an ``Observable``. The ``Observable``

Author: I changed to "An observable that returns multiple results is built upon a cursor...". Let me know if that's accurate

is a high-level abstraction built on top of a cursor, or a 
mechanism that fetches results in batches. The {+driver-short+}
handles the cursor internally and exposes the operation
results through an observable interface.

Each operation returns an instance of the ``Observable[T]`` type, where the ``[T]``
parameterization corresponds to the data type that the ``Observable``

> stores. Some operations return a specialized observable that

Author: Changed "stores" to "handles"

extends the general ``Observable[T]`` type and provides additional
methods to filter and process results. These specialized observables
include the following types:

> - ``SingleObservable[T]``: Returned by operations that produce a single result

Author: Reworded - are FindObservable, DistinctObservable, etc. "specific Observable types"? Or is there a better word than "types"?

over time. To access this data, you can create an ``Observer`` instance
that subscribes to the ``Observable``. 

.. note::

   The {+driver-short+} is built upon the MongoDB Java Reactive Streams driver.
   The ``Observable`` class extends the ``Publisher`` class from Java Reactive
   Streams and includes additional convenience methods to help process results.

How to Process an Observable
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To run a MongoDB operation and process its data, you must request
the operation results from an ``Observable``. An ``Observable``

> that returns multiple results is built on top of a cursor, or a 

Reviewer 2: It's better, but thinking about it more, it's both incomplete and inaccurate

# Not all Observable instances are built on top of a cursor.  Some commands that return multiple results, like `distinct` don't actually use cursors
# If we are going to say what one type of observable is built on, it feels incomplete not to say what the others are built on.

I wonder if we can make this more abstract, and just say that the Observable interface is used for operations that can return an unbounded number of items (including 0, for some operations, like `FindObservable#first`), while SingleObservable is used for operations that return either 0 or 1 items.  And then give an example of each.

What do you think?

Author: Okay, I reworded to make this part more generic. I removed the mention of cursors and discussed the SingleObservable for 0/1 results and Observable for an unbounded number. Let me know what you think.

mechanism that fetches results in batches. The {+driver-short+}
handles the cursor internally and exposes the operation
results through an observable interface.

Each operation returns an instance of the ``Observable[T]`` type, where the ``[T]``
parameterization corresponds to the data type that the ``Observable``
handles. Operations that produce a single result, such as the ``findOne()``
method, return an instance of ``SingleObservable[T]``. Some operations return
specific ``Observable`` types that provide additional methods to filter and

> process results. The following types allow you to chain operation-specific

Reviewer 2: ```suggestion
process results before subscribing to them. The following types allow you to chain operation-specific
```

Reviewer 2: Can you apply this suggestion?

Author: made this change in the last commit!

> process results. The following types allow you to chain operation-specific

Reviewer 2: There are more than just these three.  Do you want to list them all, or say that these are illustrative?

Author: I don't mean it to be a complete list. I'll make that clear when introducing it
